File: aws-s3/src/controller/ImageController.ts
import { Controller, Get, Param, StreamableFile } from '@nestjs/common';
import { ImageService } from '../service/ImageService';

@Controller()
export class ImageController {
  constructor(private imageService: ImageService) {}

  @Get('image/:userId/:filename')
  async getImage(@Param('userId') userId: string, @Param('filename') filename: string): Promise<StreamableFile> {
    const stream = await this.imageService.getImageStream(userId, filename);

    return new StreamableFile(stream);
  }
}

File: aws-s3/spring-boot/src/main/java/com/example/controller/ImageController.java
package com.example.controller;

import com.example.service.ImageService;
import lombok.RequiredArgsConstructor;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpHeaders;
import org.springframework.http.ResponseEntity;
import org.springframework.web.bind.annotation.*;

@RestController
@RequiredArgsConstructor
public class ImageController {

    private final ImageService imageService;

    @GetMapping("/image/{userId}/{filename}")
    @ResponseBody
    public ResponseEntity<Resource> getImage(
            @PathVariable String userId, @PathVariable String filename) {
        Resource file = imageService.getImage(userId, filename);

        if (file == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok()
                .header(
                        HttpHeaders.CONTENT_DISPOSITION,
                        "attachment; filename=\"" + file.getFilename() + "\"")
                .body(file);
    }
}
